design_notes/XDN03 - xlang Type System.md

```diff
@@ -15,3 +15,444 @@ status: draft
 This design note describes the xlang's [type system](https://en.wikipedia.org/wiki/Type_system).
 The type system is designed to interoperate with a variety of object-oriented programming
 languages in a natural and familiar way.
+
+## Philosophy
+
+xlang is designed to interoperate with many different object-oriented programming languages. Some
+of the languages xlang is considering at this time include (but is not limited to):
+
+- [C++11](https://isocpp.org)
+- [Java](https://www.oracle.com/java) and [Kotlin](https://kotlinlang.org)
+- [JavaScript](https://developer.mozilla.org/en-US/docs/Web/JavaScript)  
+- [.NET](https://dotnet.microsoft.com), in particular
+  [C#](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/) and
+  [Visual Basic](https://docs.microsoft.com/en-us/dotnet/visual-basic/language-reference/)
+- [Objective-C](http://en.wikipedia.org/wiki/Objective-C) and [Swift](https://swift.org)
+- [Python](https://python.org)
+- [TypeScript](https://www.typescriptlang.org/)
+
+These languages represent a wide variety of philosophical differences in type system design.
+Strong vs. weak types. Compiled vs interpreted languages. Static vs dynamic typing. Gradual typing.
+Designing a single type system that can work is such varied environments inevitably leads to
+compromises that an individual language designer would never face.
+
+## Type Categories
+
+### Namespaces
+
+A namespace is a naming scope used to organize code and avoid naming collisions. All type
+categories in the xlang type system (enumerations, structs, delegates, interfaces, and runtime classes)
+except for fundamental types live in a namespace.
+
+Namespaces may contain other namespaces.
+
+xlang namespaces and type names are case preserving but insensitive. This means that you cannot
+have namespaces or type names that vary only by case. For example: you cannot have `Foo.SomeType`
+and `foo.SomeType` nor can you have `Foo.SomeType` and `Foo.someType`. However, you can have
+`Foo.SomeType` and `foo.AnotherType`. In this case `SomeType` and `AnotherType` would be considered
+as belonging to the same namespace.
+
+### Fundamental Types
+
+The xlang Type system includes a core set of built-in primitive types.
+
+xlang Type | Type Description
+---------- | ----------------------------------------------------------
+Int8       | 8 bit signed integer
+Int16      | 16 bit signed integer
+Int32      | 32 bit signed integer
+Int64      | 64 bit signed integer
+UInt8      | 8 bit unsigned integer
+UInt16     | 16 bit unsigned integer
+UInt32     | 32 bit unsigned integer
+UInt64     | 64 bit unsigned integer
+Single     | 32-bit IEEE 754 floating point number
+Double     | 64-bit IEEE 754 floating point number
+Char16     | 16-bit non-numeric value representing a UTF-16 code unit
+Boolean    | 8-bit Boolean value
+String     | immutable sequence of Char16s used to represent text
```

Comment:
Many projections choose to make the String object a full fledged object, which introduces the possibility to have things like strings in heterogeneous "object maps" etc. Would Xlang consider making its string type an xlangobject?

Reply:
I think strings need to be an fundamental type, but I could see a specific language projection customizing how strings are projected so they work better with that language's type system